<a href="https://colab.research.google.com/github/murilosantanaa/Carreira-Data-Science---Alura/blob/main/MODULO-02-Python-para-Dados/Hora_da_pratica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [ ]:
print ('Escola de Dados da Alura!')

Escola de Dados da Alura!


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [ ]:
print ('Murilo')
print ('Santana')

Murilo
Santana


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [ ]:
print('M')
print('U')
print('R')
print('I')
print('L')
print('O')

M
U
R
I
L
O


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [ ]:
print (14,'julho',1994)

14 julho 1994


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
print(2025)

2025


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [ ]:
nome = input('Digite o seu nome :')
print (f"Olá, {nome} !")

Digite o seu nome :Murilo Santana
Olá, Murilo Santana !


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [ ]:
nome = input ('Digite o seu nome :')
idade = int(input ('Digite a sua idade :'))
print (f"Olá {nome},você tem {idade} anos.")

Digite o seu nome :Murilo Santana
Digite a sua idade :31
Olá Murilo Santana,você tem 31 anos.


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [ ]:
nome = input ('Digite o seu nome :')
idade = int (input ('Digite a sua idade :'))
altura = float (input ('Digite a sua altura :'))
print (f'Olá, {nome}, você tem {idade} anos e mede {altura} metros!')

Digite o seu nome :Murilo Santana
Digite a sua idade :31
Digite a sua altura :1.81
Olá, Murilo Santana, você tem 31 anos e mede 1.81 metros!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
numero1 = int(input('Digite o primeiro número :'))
numero2 = int(input('Digite o segundo número:'))

soma = numero1 + numero2

print (f'A soma dos números é = {soma}.')

Digite o primeiro número :20
Digite o segundo número:30
A soma dos números é = 50.


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
numero1 = int(input('Digite o primeiro número :'))
numero2 = int(input('Digite o segundo número:'))
numero3 = int(input('Digite o terceiro número: '))

soma = numero1 + numero2 + numero3

print (f'O valor da soma dos números é = {soma} ')

Digite o primeiro número :20
Digite o segundo número:100
Digite o terceiro número: 2
O valor da soma dos números é = 122 


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [ ]:
numero1 = int(input('Digite o primeiro número :'))
numero2 = int(input('Digite o segundo número :'))

subtracao = numero1 - numero2

print (f'A subtração dos dois números é = {subtracao}')

Digite o primeiro número :100
Digite o segundo número :120
A subtração dos dois números é = -20


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [ ]:
numero1 = int(input('Digite o primeiro número :'))
numero2 = int(input('Digite o segundo número :'))

multiplicacao = numero1 * numero2

print(f'O valor da multiplicação dos números é = {multiplicacao}')

Digite o primeiro número :10
Digite o segundo número :50
O valor da multiplicação dos números é = 500


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
numerador = int(input('Digite o numerador :'))
denominador = int(input('Digite o denominador :'))

divisao = numerador // denominador

print(f'O valor da divisão é = {divisao}')

Digite o numerador :100
Digite o denominador :50
O valor da divisão é = 2


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
operador = int(input('Digite o operador: '))
potencia = int(input('Digite a potencia: '))

exponenciacao = operador ** potencia

print(f'O valor da exponenciacao é {exponenciacao}')

Digite o operador: 10
Digite a potencia: 2
O valor da exponenciacao é 100


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
numerador = int(input('Digite o numerador :'))
denominador = int(input('Digite o denominador :'))

divisao = int(numerador / denominador)

print(f'O valor da divisão é = {divisao}')

Digite o numerador :100
Digite o denominador :20
O valor da divisão é = 5


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
numerador = int(input('Digite o numerador :'))
denominador = int(input('Digite o denominador :'))

resto = numerador % denominador

print (f'O resto da divisao de {numerador} e {denominador} é = {resto}. ')

Digite o numerador :100
Digite o denominador :5
O resto da divisao de 100 e 5 é = 0. 


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [ ]:
n1 = float(input('1ª nota: '))
n2 = float(input('2ª nota: '))
n3 = float(input('3ª nota: '))

media = n1 + n2 + n3 // 3

print(f'A média das notas {n1},{n2},{n3} é  = {media}')

1ª nota: 20
2ª nota: 50
3ª nota: 40
A média das notas 20.0,50.0,40.0 é  = 83.0


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
n1 = 5
n2 = 12
n3 = 20
n4 = 15

media = (n1 * 1 + n2 * 2 + n3 * 3 + n4 * 4) / (1 + 2 + 3 + 4)

print(f'A média entre os números é = {media}')

A média entre os números é = 14.9


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [ ]:
frase = 'Vai dar tudo certo!'
print (frase)

Vai dar tudo certo!


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [ ]:
frase = input('Digite uma frase:')
print(frase)

Digite uma frase:Fé em Deus ! 
Fé em Deus ! 


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [ ]:
frase = input('Digite uma frase: ')
print(frase.upper())

Digite uma frase: Glória ! Glória!  
GLÓRIA ! GLÓRIA!  


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [ ]:
frase = input ('Digite uma frase: ')
print (frase.lower())

Digite uma frase: AMÉM , AMÉM !! 
amém , amém !! 


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [ ]:
frase = input ('Digite uma frase: ')
print (frase.strip())

Digite uma frase:                     Glória       Glória , Aleluiaa !
Glória       Glória , Aleluiaa !


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [ ]:
frase = input ('Digite uma frase: ')
print (frase.strip())

Digite uma frase: Deus sabe de todas as coisas ! 
Deus sabe de todas as coisas !


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [ ]:
frase = input ('Digite uma frase: ')
print (frase.strip().lower())

Digite uma frase:                       O NEYMAR É FERA ! 
o neymar é fera !


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [ ]:
frase = input ('Digite uma frase : ')
print (frase.replace ('e','f') )

Digite uma frase : Verdade , Mentira 
Vfrdadf , Mfntira 


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [ ]:
frase = input ('Digite uma frase : ')
print (frase.replace ('a','@'))

Digite uma frase : Maldade é massa
M@ld@de é m@ss@


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [ ]:
frase = input ('Digite uma frase : ')
print (frase.replace ('s','$'))

Digite uma frase : Dinheiros ssssssssssss
Dinheiro$ $$$$$$$$$$$$


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [ ]:
n1 = int(input('Digite um número :'))
n2 = int(input('Digite outro número : '))

if n1 > n2:
  print(f'{n1} é maior que {n2}')
else:
  print(f'{n2} é maior que {n1}')

Digite um número :100
Digite outro número : 30
100 é maior que 30


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [ ]:
percentualcrescimento = float(input ('Digite a porcentagem de crescimento de produção :'))

if percentualcrescimento > 0:
  print(f'O percentual de {percentualcrescimento} % é positivo !')
else:
  print(f'O percentual de {percentualcrescimento} % é negativo !')

Digite a porcentagem de crescimento de produção :20
O percentual de 20.0 % é positivo !


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
letra = input('Digite uma letra :').lower()

vogais = 'a,e,i,o,u'

if letra in vogais:
  print(f'A letra "{letra}" é uma vogal.')
else:
  print (f'A letra "{letra}" é uma consoante.')

Digite uma letra :Z
A letra "z" é uma consoante.


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:
valor1 = float(input('Digite o valor médio do carro em 2023: '))
valor2 = float(input('Digite o valor médio do carro em 2024: '))
valor3 = float(input('Digite o valor médio do carro em 2025: '))

# Inicializa maior e menor valor com o primeiro valor(Ponto de partida atribuindo o valor1 ao maior e menor valor para comparar em cascata.)
maior_valor = valor1
menor_valor = valor1

# Compara com o segundo valor
if valor2 > maior_valor:
  maior_valor = valor2
if valor2 < menor_valor:
  menor_valor = valor2

# Compara com o terceiro valor
if valor3 > maior_valor:
  maior_valor = valor3
if valor3 < menor_valor:
  menor_valor = valor3

print(f'O valor mais alto é R${maior_valor}')
print(f'O valor mais baixo é R${menor_valor}')

Digite o valor médio do carro em 2023: 15000
Digite o valor médio do carro em 2024: 20000
Digite o valor médio do carro em 2025: 9000
O valor mais alto é R$20000.0
O valor mais baixo é R$9000.0


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [ ]:
produto1 = float(input('Informe o preço do produto 1 : '))
produto2 = float(input('Informe o preço do produto 2 : '))
produto3 = float(input('Informe o preço do produto 3 : '))

if produto1 <= produto2 and produto1 <= produto3:
  print(f'O produto 1 é o mais barato.')
elif produto2 <= produto1 and produto2 <= produto3:
  print(f'O produto 2 é o mais barato.')
else:
  print(f'O produto 3 é o mais barato')

Informe o preço do produto 1 : 23
Informe o preço do produto 2 : 5
Informe o preço do produto 3 : 200
O produto 2 é o mais barato.


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [ ]:
numero1 = int(input('Informe o primeiro número: '))
numero2 = int(input('Informe o segundo número: '))
numero3 = int(input('Informe o terceiro número: '))

# Verifica qual é o maior número e então compara os outros dois
if numero1 >= numero2 and numero1 >= numero3:
  if numero2 >= numero3:
    print(f'{numero1}, {numero2}, {numero3}')
  else:
    print(f'{numero1}, {numero3}, {numero2}')
elif numero2 >= numero1 and numero2 >= numero3:
  if numero1 >= numero3:
    print(f'{numero2}, {numero1}, {numero3}')
  else:
    print(f'{numero2}, {numero3}, {numero1}')
else:
  if numero1 >= numero2:
    print(f'{numero3}, {numero1}, {numero2}')
  else:
    print(f'{numero3}, {numero2}, {numero1}')

Informe o primeiro número: 100
Informe o segundo número: 200
Informe o terceiro número: 50
200, 100, 50


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [ ]:
print('========== T U R N O S =========')
print()
print('1 - Manhã')
print('2 - Tarde')
print('3 - Noite')

print() # Adiciona uma linha em branco aqui

periodo = int(input('Digite o número correspondente ao seu turno: '))

if periodo == 1:
  print('Bom dia!')
elif periodo == 2:
  print('Boa tarde!')
elif periodo == 3:
  print('Boa noite!')
else:
  print('Valor Inválido!')

========== T U R N O S =========

1 - Manhã
2 - Tarde
3 - Noite

Digite o número correspondente ao seu turno: 0
Valor Inválido!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [ ]:
numero = int(input('Digite um número qualquer: '))

if numero % 2 == 0:
  print ('O numero é par.')
else:
  print ('O numéro é impar.')

Digite um número qualquer: 5
O numéro é impar.


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [ ]:
try:
  numero = float(input('Digite um número: '))

  # Verifica se o número float é um inteiro usando o método is_integer()
  if numero.is_integer():
    # Para exibir o número original como string, podemos usar str(int(numero))
    # ou o input original, mas o enunciado pede 'o número {numero_str}',
    # então adaptando para mostrar o número digitado sem ser a variável string original
    print(f'O número {numero} é um número inteiro.')
  else:
    print(f'O número {numero} é um número decimal.')
except ValueError:
  print('Entrada inválida. Por favor, digite um número válido.')

Digite um número: cinco
Entrada inválida. Por favor, digite um número válido.


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [ ]:
numero1 = float(input('Digite o primeiro número: '))
numero2 = float(input('Digite o segundo número: '))

print ()

print('Qual operação você quer realizar?')
print()
print (' ======= OPERAÇÕES ======= ')
print ()
print ('1 - Soma')
print ('2 - Subtração')
print ('3 - Divisão')
print ('4 - Multiplicação')
print ()
operacao = int(input('Digite a operação desejada: '))

soma = numero1 + numero2
subtracao = numero1 - numero2
divisao = numero1 / numero2
multiplicacao = numero1 * numero2

if operacao == 1:
  print(f'A soma dos números é = {soma}')
elif operacao == 2:
  print(f'A subtração dos números é = {subtracao}')
elif operacao == 3:
  print(f'A divisão dos números é = {divisao}')
elif operacao == 4:
  print(f'A multiplicação dos números é = {multiplicacao}')
else:
  print('Operação inválida.')

print()

if numero1 % 2 == 0:
    print(f'O número {numero1} é par.')
else:
  print(f'O número {numero1} é impar.')

if numero1 > 0:
  print(f'O número {numero1} é positivo.')
else:
  print(f'O número {numero1} é negativo.')

if numero1 == int(numero1):
  print(f'O número {numero1} é inteiro.')
else:
  print(f'O número {numero1} é decimal.')


print ()
## divisão de descrição entre os números.

if numero2 % 2 == 0:
  print(f'O número {numero2} é par.')
else:
  print(f'O número {numero2} é impar.')

if numero2 > 0:
  print(f'O número {numero2} é positivo.')
else:
  print(f'O número {numero2} é negativo.')

if numero2 == int(numero2):
  print(f'O número  {numero2} é inteiro.')
else:
  print(f'O número {numero2} é decimal.')

print ()



Digite o primeiro número: 800
Digite o segundo número: 900

Qual operação você quer realizar?

 ======= OPERAÇÕES ======= 

1 - Soma
2 - Subtração
3 - Divisão
4 - Multiplicação

Digite a operação desejada: 2
A subtração dos números é = -100.0

O número 800.0 é par.
O número 800.0 é positivo.
O número 800.0 é inteiro.

O número 900.0 é par.
O número 900.0 é positivo.
O número  900.0 é inteiro.



#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [ ]:
numero1 = int(input('Digite o valor do primeiro número : '))
numero2 = int(input('Digite o valor do segundo número : '))
numero3 = int(input('Digite o valor do terceiro número : '))

# Verifica a condição da desigualdade triangular
if (numero1 + numero2 > numero3) and (numero1 + numero3 > numero2) and (numero2 + numero3 > numero1):
  if numero1 == numero2 == numero3:
    print ('O triângulo é equilátero.')
  elif numero1 == numero2 or numero1 == numero3 or numero3 == numero2:
    print ('O triângulo é isósceles.')
  else:
    print ('O triângulo é escaleno.')
else:
  print ('Não é um triângulo.')

Digite o valor do primeiro número : 1
Digite o valor do segundo número : 2
Digite o valor do terceiro número : 10
Não é um triângulo.


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [ ]:
print ('========= TIPOS DE COMBUSTIVEL ===========')
print('1 - ETANOL')
print('2 - DIESEL')
print()

litrodiesel = 2
litroetanol = 1.7

combustivel = float(input('Digite o tipo de combustivel escolhido : '))
quantidade = float(input('Digite a quantidade desejada de combustível em litros :'))

valortotaletanol = litroetanol * quantidade
valortotaldiesel = litrodiesel * quantidade

if combustivel == 1 and quantidade <= 15:
  print (f'Valor total : R$ {(valortotaletanol)-(valortotaletanol * 0.02)}')
elif combustivel == 1 and quantidade > 15:
  print (f'Valor total : R$ {(valortotaletanol)-(valortotaletanol * 0.04)}')
elif combustivel == 2 and quantidade <= 15:
  print (f'Valor total : R$ {(valortotaldiesel)-(valortotaldiesel * 0.03)}')
elif combustivel == 2 and quantidade > 15:
  print (f'Valor total : R$ {(valortotaldiesel)-(valortotaldiesel * 0.05)}')
else:
  print ('Entradas incorretas.')



========= TIPOS DE COMBUSTIVEL ===========
1 - ETANOL
2 - DIESEL

Digite o tipo de combustivel escolhido : 2
Digite a quantidade desejada de combustível em litros :60
Valor total : R$ 114.0


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [ ]:
# Coleta os dados de vendas anuais
vendas_2022 = float(input('Digite o valor de vendas de 2022: '))
vendas_2023 = float(input('Digite o valor de vendas de 2023: '))

# Calcula a variação percentual
if vendas_2022 == 0:
    print('Não é possível calcular a variação percentual com vendas de 2022 iguais a zero.')
else:
    variacao_percentual = ((vendas_2023 - vendas_2022) / vendas_2022) * 100
    print(f'Variação percentual: {variacao_percentual:.2f}%')

    # Envia sugestões baseadas na variação
    if variacao_percentual > 20:
        print('Sugestão: Bonificação para o time de vendas.')
    elif 2 <= variacao_percentual <= 20:
        print('Sugestão: Pequena bonificação para o time de vendas.')
    elif -10 <= variacao_percentual < 2:
        print('Sugestão: Planejamento de políticas de incentivo às vendas.')
    else: # variacao_percentual < -10
        print('Sugestão: Corte de gastos.')


Digite o valor de vendas de 2022: 20000000
Digite o valor de vendas de 2023: 2222220
Variação percentual: -88.89%
Sugestão: Corte de gastos.


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [ ]:
num1 = int(input("Digite o primeiro número inteiro : "))
num2 = int(input("Digite o segundo número inteiro : "))

menor = min(num1,num2)
maior = max(num1,num2)

for numero in range(menor,maior) :
  print (numero)

Digite o primeiro número inteiro : 100
Digite o segundo número inteiro : 3000
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277
278
279
280
281
282
283
284
285
286
287
288
289
290
291
292
293
294
295
296
297
298
299
300
301
302
303
304
305
306
307
308
309
310
311
312
313
314
315
316
317
318
319
320
321
322
323
324
325
326
327
328
329
33

#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [ ]:
dias = 0

colA = 4
colB = 10

while colA < colB:
  colA *= 1.03
  colB *= 1.015
  dias += 1

print(f'Dia : {dias}')
print(f'Colonia A : {colA:.2f}')
print(f'Colonia B : {colB:.2f}')
print()



Dia : 63
Colonia A : 25.75
Colonia B : 25.55



#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [ ]:
# Vamos repetir o processo 15 vezes (uma para cada usuário)
for i in range(15):

    # Pede a nota inicial
    nota = float(input(f"Digite a nota do usuário {i+1}: "))

    # VALIDAÇÃO: Enquanto a nota for menor que 0 OU maior que 5...
    while (nota < 0) or (nota > 5):
        print("❌ Nota Inválida! A nota deve ser entre 0 e 5.")
        # Pede a nota novamente (e atualiza a variável 'nota')
        nota = float(input(f"Tente novamente a nota do usuário {i+1}: "))

    # Se saiu do while, significa que a nota é válida
    print("✅ Nota Válida computada.")

print("Fim das avaliações!")

Digite a nota do usuário 1: 9
❌ Nota Inválida! A nota deve ser entre 0 e 5.
Tente novamente a nota do usuário 1: 5
✅ Nota Válida computada.
Digite a nota do usuário 2: 2
✅ Nota Válida computada.
Digite a nota do usuário 3: 3
✅ Nota Válida computada.
Digite a nota do usuário 4: 5
✅ Nota Válida computada.
Digite a nota do usuário 5: 4
✅ Nota Válida computada.
Digite a nota do usuário 6: 8
❌ Nota Inválida! A nota deve ser entre 0 e 5.
Tente novamente a nota do usuário 6: 9
❌ Nota Inválida! A nota deve ser entre 0 e 5.
Tente novamente a nota do usuário 6: 2
✅ Nota Válida computada.
Digite a nota do usuário 7: 4
✅ Nota Válida computada.
Digite a nota do usuário 8: 5
✅ Nota Válida computada.
Digite a nota do usuário 9: 1
✅ Nota Válida computada.
Digite a nota do usuário 10: 2
✅ Nota Válida computada.
Digite a nota do usuário 11: 2.5
✅ Nota Válida computada.
Digite a nota do usuário 12: 4
✅ Nota Válida computada.
Digite a nota do usuário 13: 6
❌ Nota Inválida! A nota deve ser entre 0 e 5.
Ten

#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [ ]:
somatemperatura = 0
quantidade = 0

temperatura = float(input('Digite o valor da temperatura em Celsius ou digite -273 para sair. '))

while temperatura != -273:

  somatemperatura += temperatura
  quantidade += 1

  temperatura = float(input('Digite o valor da temperatura em Celsius: '))


if quantidade > 0:

  media = somatemperatura / quantidade

  print(f'A média das temperaturas é de {media} C°')



Digite o valor da temperatura em Celsius : 100
Ou digite -273 para sair.
Digite o valor da temperatura em Celsius: 300
Digite o valor da temperatura em Celsius: 500
Digite o valor da temperatura em Celsius: 400
Digite o valor da temperatura em Celsius: 500
Digite o valor da temperatura em Celsius: 600
Digite o valor da temperatura em Celsius: 200
Digite o valor da temperatura em Celsius: 5
Digite o valor da temperatura em Celsius: 10
Digite o valor da temperatura em Celsius: 20
Digite o valor da temperatura em Celsius: 30
Digite o valor da temperatura em Celsius: 20
Digite o valor da temperatura em Celsius: 10
Digite o valor da temperatura em Celsius: 50
Digite o valor da temperatura em Celsius: 60
Digite o valor da temperatura em Celsius: 50
Digite o valor da temperatura em Celsius: 40
Digite o valor da temperatura em Celsius: 50
Digite o valor da temperatura em Celsius: 20
Digite o valor da temperatura em Celsius: 30
Digite o valor da temperatura em Celsius: 20
Digite o valor da temp

#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [ ]:
# 1. Entrada de dados
numero = int(input("Digite um número inteiro: "))

# 2. Peças do quebra-cabeça
valoracumulado = 1
contador = numero

print(f"\n⚙️ Iniciando o cálculo de {numero}! (fatorial)")
print("-" * 40)

# 3. O Laço de Repetição
while contador > 1:
    # Guardamos o valor antes da conta apenas para exibir no print depois
    valor_anterior = valoracumulado

    # Realizamos a multiplicação
    valoracumulado = valoracumulado * contador

    # EXIBIÇÃO DO RACIOCÍNIO:
    print(f"🔄 No degrau {contador}:")
    print(f"   Conta: {valor_anterior} (acumulado) x {contador} (contador)")
    print(f"   Novo valor na caixa: {valoracumulado}")
    print("-" * 40)

    # Diminuímos o contador para descer um degrau
    contador = contador - 1

# 4. Exibição do resultado final
print(f"✅ Fim do cálculo! O fatorial de {numero} é: {valoracumulado}")


Digite um número inteiro: 5

⚙️ Iniciando o cálculo de 5! (fatorial)
----------------------------------------
🔄 No degrau 5:
   Conta: 1 (acumulado) x 5 (contador)
   Novo valor na caixa: 5
----------------------------------------
🔄 No degrau 4:
   Conta: 5 (acumulado) x 4 (contador)
   Novo valor na caixa: 20
----------------------------------------
🔄 No degrau 3:
   Conta: 20 (acumulado) x 3 (contador)
   Novo valor na caixa: 60
----------------------------------------
🔄 No degrau 2:
   Conta: 60 (acumulado) x 2 (contador)
   Novo valor na caixa: 120
----------------------------------------
✅ Fim do cálculo! O fatorial de 5 é: 120


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [ ]:
# 1. Pedimos o número que o usuário quer multiplicar
numero = int(input("Digite um número inteiro para ver a tabuada: "))

print(f"\n📊 Tabuada do {numero}:")
print("-" * 20)  # Uma linha de separação para ficar bonito

# 2. O Laço 'for'
# Usamos range(1, 11) porque o Python exclui o último número.
# Então ele vai contar: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10.
for i in range(1, 11):

    # 3. Calculamos o resultado da linha atual
    resultado = numero * i

    # 4. Exibimos no formato "X x Y = Z"
    print(f"{numero} x {i} = {resultado}")

print("-" * 20)




Digite um número inteiro para ver a tabuada: 10

📊 Tabuada do 10:
--------------------
10 x 1 = 10
10 x 2 = 20
10 x 3 = 30
10 x 4 = 40
10 x 5 = 50
10 x 6 = 60
10 x 7 = 70
10 x 8 = 80
10 x 9 = 90
10 x 10 = 100
--------------------


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

In [ ]:
numero = int(input('Digite um número inteiro: '))

if numero <= 1:
  print(f'O número {numero} não é primo.')
elif numero == 2:
  print(f'O número {numero} é primo.')
else:
  is_primo = True
  divisor = 2
  # Laço de repetição mais simples, verifica divisores até o número - 1
  while divisor < numero:
    if numero % divisor == 0:
      is_primo = False
      break # Sai do laço se encontrar um divisor
    divisor += 1

  if is_primo:
    print(f'O número {numero} é primo.')
  else:
    print(f'O número {numero} não é primo.')

#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

In [ ]:
intervalo_0_25 = 0
intervalo_26_50 = 0
intervalo_51_75 = 0
intervalo_76_100 = 0

print('Digite as idades dos clientes. Digite um número negativo para encerrar.')

while True:
  idade = int(input('Digite a idade: '))

  if idade < 0:
    break
  elif idade <= 25:
    intervalo_0_25 += 1
  elif idade <= 50:
    intervalo_26_50 += 1
  elif idade <= 75:
    intervalo_51_75 += 1
  elif idade <= 100:
    intervalo_76_100 += 1
  else:
    print('Idade fora dos intervalos considerados (0-100), não será contabilizada.')

print('\n--- Distribuição de Idades ---')
print(f'[0-25]: {intervalo_0_25} clientes')
print(f'[26-50]: {intervalo_26_50} clientes')
print(f'[51-75]: {intervalo_51_75} clientes')
print(f'[76-100]: {intervalo_76_100} clientes')

Digite as idades dos clientes. Digite um número negativo para encerrar.
Digite a idade: 10
Digite a idade: 20
Digite a idade: 50
Digite a idade: 50
Digite a idade: 40
Digite a idade: 50
Digite a idade: 60
Digite a idade: 50
Digite a idade: 80
Digite a idade: 2
Digite a idade: 3
Digite a idade: 5
Digite a idade: 65
Digite a idade: 8
Digite a idade: 54
Digite a idade: 9
Digite a idade: 6
Digite a idade: 5
Digite a idade: 2
Digite a idade: 3
Digite a idade: 5
Digite a idade: 64
Digite a idade: 52
Digite a idade: 52
Digite a idade: 35
Digite a idade: 85
Digite a idade: 13
Digite a idade: 12
Digite a idade: -5

--- Distribuição de Idades ---
[0-25]: 14 clientes
[26-50]: 6 clientes
[51-75]: 6 clientes
[76-100]: 2 clientes


#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

In [ ]:
num_funcionarios = 20

votos_candidato1 = 0
votos_candidato2 = 0
votos_candidato3 = 0
votos_candidato4 = 0
votos_nulos = 0
votos_em_branco = 0

print('--- Eleição para Gerência ---')
print('Candidatos: 1, 2, 3, 4')
print('Opções: 5 (Nulo), 6 (Em Branco)')

for i in range(num_funcionarios):
    voto = 0 # Inicializa voto para entrar no loop
    while not (1 <= voto <= 6):
        voto_str = input(f'Funcionário {i + 1}, digite seu voto (1-6): ')
        # A validação de tipo de entrada não usa try-except, assume que é um número.
        # Se o usuário digitar algo não numérico, o programa pode falhar.
        if voto_str.isdigit(): # Verifica se a entrada é composta apenas por dígitos
            voto = int(voto_str)
            if not (1 <= voto <= 6):
                print('Voto inválido. Por favor, digite um número entre 1 e 6.')
        else:
            print('Entrada inválida. Por favor, digite um número inteiro.')
            voto = 0 # Reinicia voto para manter o loop

    if voto == 1:
        votos_candidato1 += 1
    elif voto == 2:
        votos_candidato2 += 1
    elif voto == 3:
        votos_candidato3 += 1
    elif voto == 4:
        votos_candidato4 += 1
    elif voto == 5:
        votos_nulos += 1
    else: # voto == 6
        votos_em_branco += 1

total_votos = num_funcionarios

print('\n--- Resultados da Eleição ---')
print(f'Candidato 1: {votos_candidato1} votos')
print(f'Candidato 2: {votos_candidato2} votos')
print(f'Candidato 3: {votos_candidato3} votos')
print(f'Candidato 4: {votos_candidato4} votos')
print(f'Votos Nulos: {votos_nulos} votos')
print(f'Votos em Branco: {votos_em_branco} votos')

# Cálculo das porcentagens
porcentagem_nulos = (votos_nulos / total_votos) * 100 if total_votos > 0 else 0
porcentagem_em_branco = (votos_em_branco / total_votos) * 100 if total_votos > 0 else 0

print(f'Porcentagem de votos nulos: {porcentagem_nulos:.2f}%')
print(f'Porcentagem de votos em branco: {porcentagem_em_branco:.2f}%')

# Determinar o vencedor
vencedor = None
max_votos = -1

# Criando uma lista simples de candidatos e seus votos para facilitar a comparação
candidatos_votos = [
    (1, votos_candidato1),
    (2, votos_candidato2),
    (3, votos_candidato3),
    (4, votos_candidato4)
]

# Encontra o máximo de votos entre os candidatos
# Usa uma flag para detectar empates
possiveis_vencedores = []

for cand_num, votos_atual in candidatos_votos:
    if votos_atual > max_votos:
        max_votos = votos_atual
        possiveis_vencedores = [cand_num] # Novo maior, reseta a lista
    elif votos_atual == max_votos and max_votos > 0: # Adiciona em caso de empate
        possiveis_vencedores.append(cand_num)

if len(possiveis_vencedores) > 1:
    print(f'\nResultado: Empate entre os Candidatos {', '.join(map(str, possiveis_vencedores))} com {max_votos} votos!')
elif len(possiveis_vencedores) == 1:
    print(f'\nO vencedor da eleição é o Candidato {possiveis_vencedores[0]} com {max_votos} votos!')
else:
    print('\nNão houve votos válidos para nenhum candidato.')

--- Eleição para Gerência ---
Candidatos: 1, 2, 3, 4
Opções: 5 (Nulo), 6 (Em Branco)


KeyboardInterrupt: Interrupted by user

## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

In [9]:
valorgasto = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]
media = sum(valorgasto) / len(valorgasto)

print(sum(valorgasto))
print(len(valorgasto))
print(f'{media:.2f}')

31442.22
10
3144.22


#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

In [16]:
compras_acima_3000 = 0
total_compras = len(valorgasto)

for gasto in valorgasto:
  if gasto > 3000:
    compras_acima_3000 += 1

print (f'Tiveram {compras_acima_3000} compras acima de R$ 3000.')

porcentagem_acima_3000 = (compras_acima_3000 / total_compras) * 100

print(f'A porcentagem de compras acima de 3000 foi de {porcentagem_acima_3000} %')



Tiveram 7 compras acima de R$ 3000.
A porcentagem de compras acima de 3000 foi de 70.0 %


#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
